nspire/main.c

```diff
@@ -131,47 +163,70 @@ STATIC char *strjoin(const char *s1, int sep_char, const char *s2) {
     return s;
 }
 
-STATIC void do_repl(void) {
-    printf("Micro Python " MICROPY_GIT_TAG " on " MICROPY_BUILD_DATE "\n");
+STATIC int do_repl(void) {
+    mp_hal_stdout_tx_str("MicroPython " MICROPY_GIT_TAG " on " MICROPY_BUILD_DATE "\n");
 
-    while(!should_exit) {
-        char *line = prompt(">>> ");
-	if(!line)
-		return;
+    // use simple readline
 
+    for (;;) {
+        char *line = prompt(">>> ");
+        if (line == NULL) {
+            // EOF
+            return 0;
+        }
         while (mp_repl_continue_with_input(line)) {
             char *line2 = prompt("... ");
-            if (!line2)
+            if (line2 == NULL) {
                 break;
-
+            }
             char *line3 = strjoin(line, '\n', line2);
             free(line);
             free(line2);
             line = line3;
         }
 
-        if(strcmp("quit", line) == 0) {
-            should_exit = true;
-        } else {
-            mp_lexer_t *lex = mp_lexer_new_from_str_len(MP_QSTR__lt_stdin_gt_, line, strlen(line), false);
-            execute_from_lexer(lex, MP_PARSE_SINGLE_INPUT, true);
+        int ret = execute_from_lexer(LEX_SRC_STR, line, MP_PARSE_SINGLE_INPUT, true);
+        if (ret & FORCED_EXIT) {
+            return ret;
```

Comment:
This should be ```return ret & ~FORCED_EXIT```.

Reply:
Actually, no, that's handled correctly in main, only values from 0-255 are supported apparently.